File: apps/api/src/otel.ts
import { basename } from "node:path";
import { logs, SeverityNumber } from "@opentelemetry/api-logs";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-http";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-http";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import {
  defaultResource,
  resourceFromAttributes,
} from "@opentelemetry/resources";
import { BatchLogRecordProcessor } from "@opentelemetry/sdk-logs";
import { PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { NodeSDK } from "@opentelemetry/sdk-node";
import {
  BatchSpanProcessor,
  type ReadableSpan,
} from "@opentelemetry/sdk-trace-base";
import { consola } from "consola";
import { version } from "../package.json" with { type: "json" };

// MUST be imported before any other module (express, pg, ioredis, …) so
// auto-instrumentation can patch them, and before "./metrics" so instruments
// bind to a real MeterProvider instead of the no-op one.
//
// What ties the three signals together:
//
//   * one resource on all of them (below), so a backend groups the spans, the
//     metrics and the logs as one process rather than three streams;
//   * trace and span ids on every log record — the consola bridge at the
//     bottom emits into the active context, so a line written inside a request
//     handler links back to that request's trace;
//   * W3C traceparent in and out, which the auto-instrumentations do by
//     default, so a trace that starts here continues into the appview and the
//     Go proxies and vice versa.
//
// The fourth link, exemplars — a trace id on one histogram sample, so a chart
// clicks through to the slow request behind a spike — is not available here.
// @opentelemetry/sdk-metrics ships the exemplar types and filters but nothing
// wires them into the aggregators, and otlp-transformer does not serialize
// them, so there is no option to turn on. Getting from a metric to a trace
// means filtering the traces by the same attribute names the metric carries;
// that is why those names are kept identical across the services.

const otlpBase = (
  process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? "http://127.0.0.1:4318"
).replace(/\/+$/, "");

const tracesUrl =
  process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT ?? `${otlpBase}/v1/traces`;
const metricsUrl =
  process.env.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT ?? `${otlpBase}/v1/metrics`;
const logsUrl =
  process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT ?? `${otlpBase}/v1/logs`;

// index.ts (Hono) and server.ts (XRPC) run as separate processes from the
// same codebase; tell them apart by entrypoint unless explicitly named.
const entrypoint = basename(process.argv[1] ?? "");
const serviceName =
  process.env.OTEL_SERVICE_NAME ??
  (entrypoint.startsWith("server") ? "rocksky-xrpc" : "rocksky-api");

// Telemetry must never carry credentials or PII: redact any auth-shaped
// attribute and strip query strings (API keys, cursors, search terms) from
// every URL-shaped attribute before export.
const SENSITIVE_KEY =
  /authorization|cookie|set-cookie|token|secret|passw|api[-_]?key|session/i;
const URL_VALUE_KEY = /^(http\.url|http\.target|url\.full)$/;

function scrubAttributes(attributes: Record<string, unknown>) {
  for (const key of Object.keys(attributes)) {
    const value = attributes[key];
    if (SENSITIVE_KEY.test(key)) {
      attributes[key] = "[REDACTED]";
    } else if (key === "url.query") {
      delete attributes[key];
    } else if (URL_VALUE_KEY.test(key) && typeof value === "string") {
      attributes[key] = value.split("?")[0];
    }
  }
}

class ScrubbingSpanProcessor extends BatchSpanProcessor {
  onEnd(span: ReadableSpan) {
    scrubAttributes(span.attributes as Record<string, unknown>);
    super.onEnd(span);
  }
}

// The same identity the Rust services get from crates/telemetry and the Go
// ones from otel/, on all three signals. `defaultResource()` first so the
// telemetry.sdk.* attributes survive; the detectors NodeSDK runs afterwards
// (env, process, host) are merged on top, so OTEL_RESOURCE_ATTRIBUTES still
// overrides anything set here.
const resource = defaultResource().merge(
  resourceFromAttributes({
    "service.version": version,
    // Every Rocksky process shares a namespace, which is what lets a backend
    // show "all of Rocksky" without naming the services by hand.
    "service.namespace": "rocksky",
    // Tells the two processes built from this codebase apart across a
    // restart — pid and start time, legible enough to match against ps and
    // journalctl without a lookup.
    "service.instance.id": `${process.pid}-${Date.now()}`,
    ...(process.env.DEPLOYMENT_ENVIRONMENT
      ? { "deployment.environment.name": process.env.DEPLOYMENT_ENVIRONMENT }
      : {}),
  }),
);

const sdk = new NodeSDK({
  serviceName,
  resource,
  spanProcessors: [
    new ScrubbingSpanProcessor(new OTLPTraceExporter({ url: tracesUrl })),
  ],
  metricReader: new PeriodicExportingMetricReader({
    exporter: new OTLPMetricExporter({ url: metricsUrl }),
    exportIntervalMillis: 10_000,
  }),
  logRecordProcessors: [
    new BatchLogRecordProcessor(new OTLPLogExporter({ url: logsUrl })),
  ],
  instrumentations: [
    getNodeAutoInstrumentations({
      // fs/net/dns produce enormous span volume for no diagnostic value here
      "@opentelemetry/instrumentation-fs": { enabled: false },
      "@opentelemetry/instrumentation-net": { enabled: false },
      "@opentelemetry/instrumentation-dns": { enabled: false },
    }),
  ],
});

sdk.start();

// Bridge consola to OTel logs so every existing consola.* call ships to the
// collector, correlated with the active span.
const CONSOLA_SEVERITY: Record<number, SeverityNumber> = {
  0: SeverityNumber.ERROR,
  1: SeverityNumber.WARN,
  2: SeverityNumber.INFO,
  3: SeverityNumber.INFO,
  4: SeverityNumber.DEBUG,
  5: SeverityNumber.TRACE,
};

function redactText(text: string): string {
  return text
    .replace(/bearer\s+[\w\-.~+/=]+/gi, "Bearer [REDACTED]")
    .replace(/eyJ[\w-]+\.[\w-]+\.[\w-]+/g, "[REDACTED]")
    .replace(/(https?:\/\/[^\s"'?]+)\?[^\s"']*/g, "$1");
}

const otelLogger = logs.getLogger(serviceName);

consola.addReporter({
  log(logObj) {
    // Telemetry must never take the process down.
    try {
      const body = logObj.args
        .map((arg) => {
          if (typeof arg === "string") return arg;
          if (arg instanceof Error) return arg.stack ?? arg.message;
          try {
            return JSON.stringify(arg);
          } catch {
            return String(arg);
          }
        })
        .join(" ");
      otelLogger.emit({
        severityNumber: CONSOLA_SEVERITY[logObj.level] ?? SeverityNumber.INFO,
        severityText: logObj.type,
        body: redactText(body),
        attributes: logObj.tag ? { "log.tag": logObj.tag } : undefined,
      });
    } catch {
      // ignore
    }
  },
});

const flushAndExit = (signal: NodeJS.Signals) => {
  sdk
    .shutdown()
    .catch(() => {})
    .finally(() => process.kill(process.pid, signal));
};
process.once("SIGTERM", () => flushAndExit("SIGTERM"));
process.once("SIGINT", () => flushAndExit("SIGINT"));

File: otel/otel.go
// Package otel wires the three OpenTelemetry signals for the small Go
// services (spotify-proxy, musicbrainz, deezer), mirroring what the Rust
// services get from rocksky-telemetry and the Node services from otel.ts:
// traces and a per-request metric from middleware, runtime metrics, and logs
// bridged from the standard library into OTLP.
//
// Configuration is the standard environment: OTEL_EXPORTER_OTLP_ENDPOINT
// (defaulting to the local collector, http://127.0.0.1:4318) and
// OTEL_EXPORTER_OTLP_HEADERS for the collector's ingest token — the same
// variable the Node services already receive, so a service run under doppler
// needs nothing extra.
//
// # What ties the three signals together
//
// The same resource is on all three, so a backend groups them as one process:
// service.name, service.version, service.namespace and service.instance.id,
// matching what crates/telemetry stamps on the Rust services and what
// apps/api/src/otel.ts stamps on the Node ones.
//
// Logs carry the trace and span id of whatever span was open when they were
// written, which is why RequestLogger logs through the *request's* context —
// a slog call with context.Background() reaches the collector unlinked.
//
// The trace itself continues across the fleet through the W3C propagator set
// below: otelecho extracts it from incoming requests, and an outgoing client
// wrapped in otelhttp.NewTransport injects it into the next hop.
//
// And exemplars — a trace id hanging off one histogram sample, so a chart
// clicks straight through to the slow request behind the spike — are the link
// the Go SDK can give and the Rust, Node and Deno ones cannot. They are on by
// default here; WithExemplarFilter below says so out loud rather than leaving
// a link that only some of the fleet has resting on a default.
package otel

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"runtime/debug"
	"strings"
	"time"

	"go.opentelemetry.io/contrib/bridges/otelslog"
	"go.opentelemetry.io/contrib/instrumentation/runtime"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/log/global"
	"go.opentelemetry.io/otel/propagation"
	sdklog "go.opentelemetry.io/otel/sdk/log"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/metric/exemplar"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// Setup installs tracer, meter and logger providers for this process and
// returns a shutdown function that flushes them. Telemetry must never take
// the service down: any exporter that cannot be built is reported on stderr
// and skipped, and the returned shutdown is safe to call regardless.
func Setup(ctx context.Context, serviceName string) func(context.Context) error {
	// The default endpoint is the local collector. Set through the standard
	// variable rather than exporter options, so anything actually configured
	// in the environment wins untouched.
	if os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT") == "" {
		os.Setenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://127.0.0.1:4318")
	}
	// The name goes through the environment the default resource reads, not
	// through resource.Merge: merging a resource carrying a semconv schema
	// URL against Default()'s own conflicts whenever the two SDK versions
	// disagree, and the silent fallback shipped every service as
	// "unknown_service:main".
	if os.Getenv("OTEL_SERVICE_NAME") == "" {
		os.Setenv("OTEL_SERVICE_NAME", serviceName)
	}
	// The rest of the identity goes the same way, for the same reason:
	// OTEL_RESOURCE_ATTRIBUTES is read by the default resource's own detector,
	// so these land on all three signals without a Merge to go wrong. Appended
	// to whatever is already set rather than replacing it — the earlier entries
	// win in the SDK's parser, so anything configured in the environment
	// overrides these defaults.
	addResourceAttributes(map[string]string{
		"service.version": buildVersion(),
		// Every Rocksky process shares a namespace, which is what lets a
		// backend show "all of Rocksky" without naming the services by hand.
		"service.namespace": "rocksky",
		// Distinguishes two processes of the same service, and survives a
		// restart as a different value so a trace can be pinned to the run
		// that produced it. Legible on purpose: pid and start time match up
		// with ps and journalctl without a lookup.
		"service.instance.id": fmt.Sprintf("%d-%d", os.Getpid(), time.Now().UnixNano()),
		"process.pid":         fmt.Sprintf("%d", os.Getpid()),
	})
	if environment := os.Getenv("DEPLOYMENT_ENVIRONMENT"); environment != "" {
		addResourceAttributes(map[string]string{"deployment.environment.name": environment})
	}
	res := resource.Default()

	var shutdowns []func(context.Context) error

	// The trace of an incoming request continues across services.
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{}, propagation.Baggage{},
	))

	if exporter, err := otlptracehttp.New(ctx); err == nil {
		provider := sdktrace.NewTracerProvider(
			sdktrace.WithBatcher(exporter),
			sdktrace.WithResource(res),
		)
		otel.SetTracerProvider(provider)
		shutdowns = append(shutdowns, provider.Shutdown)
	} else {
		slog.Error("otel: trace exporter disabled", "error", err)
	}

	if exporter, err := otlpmetrichttp.New(ctx); err == nil {
		provider := sdkmetric.NewMeterProvider(
			sdkmetric.WithReader(sdkmetric.NewPeriodicReader(
				exporter, sdkmetric.WithInterval(10*time.Second),
			)),
			sdkmetric.WithResource(res),
			// Attach the trace and span id of the sampled span a measurement
			// was taken under to the data point. This is the SDK's default;
			// naming it keeps a future default change from quietly removing
			// the only click-through from a chart to a trace that Rocksky has.
			// It is also why every Record call must be given the request's
			// context — see Metrics in echo.go.
			sdkmetric.WithExemplarFilter(exemplar.TraceBasedFilter),
		)
		otel.SetMeterProvider(provider)
		shutdowns = append(shutdowns, provider.Shutdown)
		if err := runtime.Start(); err != nil {
			slog.Error("otel: runtime metrics disabled", "error", err)
		}
	} else {
		slog.Error("otel: metric exporter disabled", "error", err)
	}

	if exporter, err := otlploghttp.New(ctx); err == nil {
		provider := sdklog.NewLoggerProvider(
			sdklog.WithProcessor(sdklog.NewBatchProcessor(exporter)),
			sdklog.WithResource(res),
		)
		global.SetLoggerProvider(provider)
		shutdowns = append(shutdowns, provider.Shutdown)

		// Everything logged through slog — and through the standard log
		// package, which slog.SetDefault reroutes — goes to both the console
		// (journalctl keeps working) and the collector.
		console := slog.NewTextHandler(os.Stderr, nil)
		bridge := otelslog.NewHandler(serviceName, otelslog.WithLoggerProvider(provider))
		slog.SetDefault(slog.New(fanout{console, bridge}))
	} else {
		slog.Error("otel: log exporter disabled", "error", err)
	}

	return func(ctx context.Context) error {
		var errs []error
		for _, shutdown := range shutdowns {
			errs = append(errs, shutdown(ctx))
		}
		return errors.Join(errs...)
	}
}

// addResourceAttributes appends to OTEL_RESOURCE_ATTRIBUTES rather than
// setting it, and appends rather than prepends: the SDK's parser keeps the
// first value it sees for a key, so anything already in the environment
// overrides what is added here.
func addResourceAttributes(attributes map[string]string) {
	pairs := make([]string, 0, len(attributes))
	// Sorted, so the variable does not churn between runs for no reason.
	for _, key := range sortedKeys(attributes) {
		pairs = append(pairs, key+"="+attributes[key])
	}
	existing := os.Getenv("OTEL_RESOURCE_ATTRIBUTES")
	if existing != "" {
		pairs = append([]string{existing}, pairs...)
	}
	os.Setenv("OTEL_RESOURCE_ATTRIBUTES", strings.Join(pairs, ","))
}

func sortedKeys(m map[string]string) []string {
	keys := make([]string, 0, len(m))
	for key := range m {
		keys = append(keys, key)
	}
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j] < keys[j-1]; j-- {
			keys[j], keys[j-1] = keys[j-1], keys[j]
		}
	}
	return keys
}

// buildVersion is the module's version as the Go toolchain recorded it. These
// services are built from a checkout rather than installed from a proxy, so
// the usual answer is the VCS revision stamped into the binary; "dev" when the
// build carried neither.
func buildVersion() string {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return "dev"
	}
	for _, setting := range info.Settings {
		if setting.Key == "vcs.revision" && setting.Value != "" {
			return setting.Value
		}
	}
	if info.Main.Version != "" && info.Main.Version != "(devel)" {
		return info.Main.Version
	}
	return "dev"
}

// fanout sends every record to all of its handlers.
type fanout []slog.Handler

func (f fanout) Enabled(ctx context.Context, level slog.Level) bool {
	for _, h := range f {
		if h.Enabled(ctx, level) {
			return true
		}
	}
	return false
}

func (f fanout) Handle(ctx context.Context, r slog.Record) error {
	var errs []error
	for _, h := range f {
		if h.Enabled(ctx, r.Level) {
			errs = append(errs, h.Handle(ctx, r.Clone()))
		}
	}
	return errors.Join(errs...)
}

func (f fanout) WithAttrs(attrs []slog.Attr) slog.Handler {
	out := make(fanout, len(f))
	for i, h := range f {
		out[i] = h.WithAttrs(attrs)
	}
	return out
}

func (f fanout) WithGroup(name string) slog.Handler {
	out := make(fanout, len(f))
	for i, h := range f {
		out[i] = h.WithGroup(name)
	}
	return out
}
